### PR TITLE
Fix tests

### PR DESCRIPTION
**Pull request by nathanhourt**
- Fix failing tests
- Move obnoxiously slow test to separate test driver

### CODE REVIEW

File: tests/tests/operation_tests.cpp [hunk 1]
```diff
@@ -145,6 +145,7 @@ BOOST_AUTO_TEST_CASE( call_order_update_test )
       GRAPHENE_REQUIRE_THROW( borrow( dan, bitusd.amount(80000), asset(0)), fc::exception );
       BOOST_TEST_MESSAGE( "attempting to claim all collateral without paying off debt" );
       GRAPHENE_REQUIRE_THROW( cover( dan, bitusd.amount(0), asset(20000)), fc::exception );
+      GRAPHENE_REQUIRE_THROW( cover( dan, bitusd.amount(0), asset(20000-1)), fc::exception );
```
Comment (abitmore) [inline tests/tests/operation_tests.cpp]: What is this new test for? Actually it will throw an exception due to triggering a blackswan which is not allowed when updating a call position. It worth testing, but better add a message above it.

**Comment by abitmore** [comment]
@nathanhourt would you please clarify what environment you're using which detected the failing tests? E.G.
* OS version
* Compiler and version
* Boost version

Thanks.

**Comment by nathanhourt** [comment]
I'm running Arch with clang 6.0.0, boost 1.67.0.

```
nathan@spiral ~/dev/bitshares-core/build/debug/tests (git)-[remotes/origin/develop] % ./chain_test -t operation_tests/call_order_update_test                                                                                                                                                                                                                                                                                                :(
Random number generator seeded to 1530975831
GRAPHENE_TESTING_GENESIS_TIMESTAMP is 1431700000
Running 1 test case...
231200ms th_a       db_management.cpp:153         open                 ] Wiping object_database due to missing or wrong version
231200ms th_a       object_database.cpp:93        wipe                 ] Wiping object database...
231200ms th_a       object_database.cpp:95        wipe                 ] Done wiping object databse.
231200ms th_a       object_database.cpp:106       open                 ] Opening object database from /tmp/graphene-tmp/17d1-b3dc-ad22-3e1d ...
231200ms th_a       object_database.cpp:111       open                 ] Done opening object database.
231248ms th_a       db_update.cpp:230             check_for_blackswan  ] *call_itr: {"id":"1.8.1","borrower":"1.2.16","collateral":20000,"debt":85000,"call_price":{"base":{"amount":16,"asset_id":"1.3.0"},"quote":{"amount":119,"asset_id":"1.3.2"}}} 
231248ms th_a       db_update.cpp:240             check_for_blackswan  ] Black Swan detected on asset USDBIT (1.3.2) at block 1: 
   Least collateralized call: 0.23529411764705882  4.25000000000000000
   Settle Price:              1.00000000000000000  1.00000000000000000
   Max:                       1.00000000000000000  1.00000000000000000

231249ms th_a       db_update.cpp:241             check_for_blackswan  ] enable_black_swan: false 
231249ms th_a       undo_database.hpp:68          ~session             ] 10 assert_exception: Assert Exception
maybe_found != nullptr: Unable to find Object 1.8.1
    {"id":"1.8.1"}
    th_a  index.hpp:111 get

    {}
    th_a  undo_database.cpp:123 undo
terminate called after throwing an instance of 'fc::assert_exception'
unknown location(0): fatal error: in "operation_tests/call_order_update_test": signal: SIGABRT (application abort requested)
../../tests/tests/operation_tests.cpp(147): last checkpoint

*** 1 failure is detected in the test module "Master Test Suite"
201 nathan@spiral ~/dev/bitshares-core/build/debug/tests (git)-[remotes/origin/develop] % ./chain_test -t operation_tests/old_call_order_update_test_after_hardfork_583                                                                                                                                                                                                                                                                         :(
Random number generator seeded to 1530975832
GRAPHENE_TESTING_GENESIS_TIMESTAMP is 1431700000
Running 1 test case...
232891ms th_a       db_management.cpp:153         open                 ] Wiping object_database due to missing or wrong version
232891ms th_a       object_database.cpp:93        wipe                 ] Wiping object database...
232891ms th_a       object_database.cpp:95        wipe                 ] Done wiping object databse.
232891ms th_a       object_database.cpp:106       open                 ] Opening object database from /tmp/graphene-tmp/7f57-bd5b-56d8-3f12 ...
232891ms th_a       object_database.cpp:111       open                 ] Done opening object database.
232935ms th_a       db_maint.cpp:897              process_hf_868_890   ] Processing hard fork core-868-890 at block 3
232935ms th_a       db_maint.cpp:981              process_hf_868_890   ] Done processing hard fork core-868-890 at block 3
232935ms th_a       db_maint.cpp:805              update_and_match_cal ] Updating all call orders for hardfork core-343 at block 3
232935ms th_a       db_maint.cpp:833              update_and_match_cal ] Done updating all call orders for hardfork core-343 at block 3
232938ms th_a       db_update.cpp:230             check_for_blackswan  ] *call_itr: {"id":"1.8.1","borrower":"1.2.16","collateral":20000,"debt":85000,"call_price":{"base":{"amount":16,"asset_id":"1.3.0"},"quote":{"amount":119,"asset_id":"1.3.2"}}} 
232938ms th_a       db_update.cpp:240             check_for_blackswan  ] Black Swan detected on asset USDBIT (1.3.2) at block 4: 
   Least collateralized call: 0.23529411764705882  4.25000000000000000
   Settle Price:              1.00000000000000000  1.00000000000000000
   Max:                       1.50000000000000000  0.66666666666666663

232938ms th_a       db_update.cpp:241             check_for_blackswan  ] enable_black_swan: false 
232939ms th_a       undo_database.hpp:68          ~session             ] 10 assert_exception: Assert Exception
maybe_found != nullptr: Unable to find Object 1.8.1
    {"id":"1.8.1"}
    th_a  index.hpp:111 get

    {}
    th_a  undo_database.cpp:123 undo
terminate called after throwing an instance of 'fc::assert_exception'
unknown location(0): fatal error: in "operation_tests/old_call_order_update_test_after_hardfork_583": signal: SIGABRT (application abort requested)
../../tests/tests/operation_tests.cpp(249): last checkpoint

*** 1 failure is detected in the test module "Master Test Suite"
```

File: libraries/chain/market_evaluator.cpp [hunk 2]
```diff
@@ -248,6 +248,11 @@ void_result call_order_update_evaluator::do_apply(const call_order_update_operat
       call_obj = &*itr;
       old_collateralization = call_obj->collateralization();
       old_debt = call_obj->debt;
+      auto new_collateral = call_obj->collateral + o.delta_collateral.amount;
+      auto new_debt = *old_debt + o.delta_debt.amount;
+
+      // Forbid zero collateral with nonzero debt (avoids divide by zero when calculating call price below)
+      FC_ASSERT(!(new_collateral == 0 && new_debt != 0), "Cannot have zero collateral and nonzero debt");
```
[the remark this answers is not in the document]
Comment (nathanhourt) [reply libraries/chain/market_evaluator.cpp]: Incorrect. The test fails, because boost captures the division by zero exception, as it occurs inside the modify functor (line 257), triggering the object to be deleted. As per boost docs, this is expected behavior:

> https://www.boost.org/doc/libs/1_67_0/libs/multi_index/doc/reference/ord_indices.html#modify
> **Exception safety:** Basic. If an exception is thrown by some user-provided operation (including mod), then the element pointed to by position is erased.

File: tests/tests/operation_tests.cpp (hunk 1 above)
Comment (nathanhourt) [reply tests/tests/operation_tests.cpp]: The underlying implementation now explicitly checks for zero collateral. This test was added to probe the edge case that collateral is not zero, but still too low to sustain the debt. I feel it is sufficiently described by the message, "attempting to claim all collateral without paying off debt"; even though it leaves one unit of collateral unclaimed, the spirit is the same.

**Comment by nathanhourt** [comment]
Oops, didn't mean to close this...

I'm not sure what ci/dockercloud is angry about, as I am apparently forbidden to view the logs.

**Comment by abitmore** [comment]
Thanks. Boost 1.66+ is known to have issues with current object database and/or undo database implementation. I believe not only these two test cases are affected. More discussions:
* https://github.com/bitshares/bitshares-fc/pull/36#issuecomment-381588452
* https://github.com/bitshares/bitshares-core/issues/852

File: tests/tests/operation_tests.cpp (hunk 1 above)
Comment (abitmore) [reply tests/tests/operation_tests.cpp]: 0 and 1 are absolutely different.

**Comment by abitmore** [comment]
Update: created a [new project](https://github.com/bitshares/bitshares-core/projects/11) to track all boost 1.66+ related issues.

**Comment by nathanhourt** [comment]
Oh, OK, so it looks like I fixed #852 then?

**Comment by nathanhourt** [comment]
Added commit to ensure we don't trip if some other modify functor ever throws in the future

File: libraries/db/include/graphene/db/generic_index.hpp [hunk 3]
```diff
@@ -67,10 +67,24 @@ namespace graphene { namespace chain {
...
+                                             exception = true;
+                                             elog("Exception while modifying object: ${e} -- object may be corrupted",
+                                                  ("e", e));
+                                          } catch (...) {
+                                             exception = true;
+                                             elog("Unknown exception while modifying object");
+                                          }
+                                       }
+                      );
+            FC_ASSERT(!exception, "Aborting for exception while modifying object");
```
Comment (abitmore) [inline libraries/db/include/graphene/db/generic_index.hpp]: Best if can re-throw the real exception rather than covering it with a new message and logging it only, so a client program will know what happened.

Comment (nathanhourt) [reply libraries/db/include/graphene/db/generic_index.hpp]: That is true. I went with the lazy option to start, because copying exceptions can be funny, but I'll play with it a bit more

**Comment by nathanhourt** [comment]
OK, now it throws the original exception rather than a placeholder.

As to #694, that looks like a non-issue to me. The original report seems to be that someone was modifying one entry's key to collide with another one? I agree with the discussion on #694, that doesn't make sense to do, and it's not clear what the expected, non-error behavior would be if such a thing were attempted.

**Comment by abitmore** [comment]
@nathanhourt the Travis building log is public but DockerCloud is not.
```
/home/travis/build/bitshares/bitshares-core/tests/tests/database_tests.cpp:140:1:   required from here
/usr/include/boost/test/test_tools.hpp:326:14: error: ambiguous overload for ‘operator<<’ (operand types are ‘std::ostream {aka std::basic_ostream<char>}’ and ‘std::nullptr_t’)
         ostr << t; // by default print the value
              ^
```

**Comment by nathanhourt** [comment]
Travis, you dunce... OK, try this...

**Comment by nathanhourt** [comment]
So I can't fix the docker test without some idea as to where it's breaking... Is there something else I can do here?

**Comment by abitmore** [comment]
@nathanhourt thanks for your contribution.

If you're asking me, I'd like to see
* the `slow_tests` change to be moved to another PR.

Other changes in this PR are mostly fine for me, except some minor issues:
* https://github.com/bitshares/bitshares-core/pull/1126#discussion_r200811164
* perhaps add a same line in the test case `old_call_order_update_test_after_hardfork_583`
* https://github.com/bitshares/bitshares-core/pull/1126#commitcomment-29629303

By the way, most core dev team members are on vacation now, I think they'll comment when came back.
